File: ApplicationCases/CrazyRockets/app/src/main/java/com/huawei/mlkit/sample/beans/World.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.mlkit.sample.beans;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.view.View;
import android.view.WindowManager;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.huawei.mlkit.sample.R;
import com.huawei.mlkit.sample.utils.SizeUtils;

public class World {
    private Context context;

    private boolean gameStart;
    private int score;
    private boolean gameOver;

    private Rockets rockets;
    private Ground ground;
    private Obstacle obstacle1;
    private Obstacle obstacle2;

    private Paint paint;
    private Paint linearPaint = new Paint();
    private int screenWidth;
    private int screenHeight;

    private Bitmap groundBitmap;
    private int groundWidth;
    private int groundHeight;

    private Bitmap obstacleBitmap;

    private int gap;
    private int distance;
    private int randomNumber;

    private RelativeLayout gameover;
    private TextView scoreText;
    private int level;
    private float magnification = 1;

    private String scoreStr;


    private long fitstTime = 0;
    private long lastTime = 0;
    private int[] color;


    public World(Context context, RelativeLayout gameover, TextView score, int level, float magnification) {
        this.context = context;
        this.gameover = gameover;
        this.scoreText = score;
        this.level = level;
        this.magnification = magnification;
        scoreStr = context.getString(R.string.score);
        color = new int[]{context.getResources().getColor(R.color.startcolor), context.getResources().getColor(R.color.endcolor)};

        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        screenWidth = wm.getDefaultDisplay().getWidth();
        screenHeight = wm.getDefaultDisplay().getHeight();

        groundBitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.ground);
        obstacleBitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.obstacle);

        groundWidth = groundBitmap.getWidth();
        groundHeight = screenHeight - groundBitmap.getHeight() / 3 * 2;

        gap = SizeUtils.dp2Px(context, 150);
        randomNumber = screenHeight - groundBitmap.getHeight() / 3 * 2 - gap - SizeUtils.dp2Px(context, 80);
        distance = screenWidth;

        paint = new Paint();
        paint.setStyle(Paint.Style.FILL_AND_STROKE);
        paint.setColor(Color.WHITE);
        paint.setTextSize(45);
        paint.setFakeBoldText(true);

        LinearGradient linearGradient = new LinearGradient(0, 0, 0, screenHeight / 5, color, new float[]{0, 1.0f}, LinearGradient.TileMode.CLAMP);
        linearPaint.setShader(linearGradient);

        start();
    }

    public void start() {
        gameStart = false;
        gameOver = false;
        ground = new Ground(groundBitmap, groundWidth, groundHeight, level);
        rockets = new Rockets(context, screenWidth, groundHeight);
        obstacle1 = new Obstacle(context, obstacleBitmap, gap, distance, randomNumber, 1, level);
        obstacle2 = new Obstacle(context, obstacleBitmap, gap, distance, randomNumber, 2, level);
        score = 0;
    }

    public void draw(Canvas canvas) {
        obstacle1.draw(canvas);
        obstacle2.draw(canvas);

        canvas.drawRect(0, 0, screenWidth, screenHeight / 5, linearPaint);

        ground.draw(canvas);
        canvas.drawText(scoreStr + score, screenWidth / 10, screenHeight / 8, paint);
        if (gameOver) {
            gameover.setVisibility(View.VISIBLE);
            scoreText.setText(score + "");
            return;
        }

        if (gameStart && !gameOver) {
            ground.step();
            rockets.draw(canvas);

            obstacle1.step();
            obstacle2.step();

            if (rockets.pass(obstacle1, obstacle2, level)) {
                lastTime = System.currentTimeMillis();
                if (lastTime - fitstTime >= 1500) {
                    score++;
                }
                fitstTime = lastTime;
            }
            if (rockets.hit(obstacle1, obstacle2)) {
                gameStart = false;
                gameOver = true;
            }
        }
        if (!gameOver) rockets.fly();

    }


    public void startGame() {
        start();
        gameStart = true;
    }

    public void setOffset(float offset) {
        if (gameStart) {
            rockets.setOffset(offset * magnification);
        }
    }
}

File: ApplicationCases/CrazyRocketKotlin/app/src/main/java/com/huawei/crazyrocketkotlin/beans/World.kt
package com.huawei.crazyrocketkotlin.beans

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.view.View
import android.view.WindowManager
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.huawei.crazyrocketkotlin.R
import com.huawei.crazyrocketkotlin.util.SizeUtils

class World (var context : Context ,gameover : RelativeLayout,score:TextView,level: Int,magnification: Float) {

    private var gameStart = false
    private var score = 0
    private var gameOver = false

    private var rockets: Rockets? = null
    private var ground: Ground? = null
    private var obstacle1: Obstacle? = null
    private var obstacle2: Obstacle? = null

    private var paint: Paint? = null
    private val linearPaint = Paint()
    private var screenWidth = 0
    private var screenHeight = 0

    private var groundBitmap: Bitmap? = null
    private var groundWidth = 0
    private var groundHeight = 0

    private var obstacleBitmap: Bitmap? = null

    private var gap = 0
    private var distance = 0
    private var randomNumber = 0

    private var gameover: RelativeLayout? = null
    private var scoreText: TextView? = null
    private var level = 0
    private var magnification = 1f

    private var scoreStr: String? = null

    private var firstTime: Long = 0
    private var lastTime: Long = 0

    private var color = intArrayOf()


    init {
        this.gameover = gameover
        this.level = level
        this.magnification = magnification
        scoreText = score
        scoreStr = context.getString(R.string.score)

        color = intArrayOf(ContextCompat.getColor(context,R.color.startcolor), ContextCompat.getColor(context,R.color.endcolor))

        val wm = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        screenWidth = wm.defaultDisplay.width
        screenHeight = wm.defaultDisplay.height

        groundBitmap = BitmapFactory.decodeResource(context.resources, R.drawable.ground)
        obstacleBitmap = BitmapFactory.decodeResource(context.resources, R.drawable.obstacle)

        groundBitmap?.let {
            groundWidth = it.width
            groundHeight = screenHeight - it.height / 3 * 2
        }

        gap = SizeUtils.dp2Px(context, 150F)
        randomNumber = screenHeight - groundBitmap!!.height / 3 * 2 - gap - SizeUtils.dp2Px(context, 80F)
        distance = screenWidth

        paint = Paint()
        paint?.apply {
            this.style = Paint.Style.FILL_AND_STROKE
            this.color = Color.WHITE
            this.textSize = 45f
            this.isFakeBoldText = true
        }


        val linearGradient = LinearGradient(0F, 0F, 0F, screenHeight / 5F, color, floatArrayOf(0f, 1.0f),
            Shader.TileMode.CLAMP
        )
        linearPaint.shader = linearGradient

        start()

    }

    fun start() {
        gameStart = false
        gameOver = false
        ground = groundBitmap?.let { Ground(it, groundWidth, groundHeight, level) }
        rockets = Rockets(context, screenWidth, groundHeight)

        obstacleBitmap?.let {
            obstacle1 = Obstacle(context, it, gap, distance, randomNumber, 1, level)
            obstacle2 = Obstacle(context, it, gap, distance, randomNumber, 2, level)
        }

        score = 0
    }

    @SuppressLint("SetTextI18n")
    fun draw(canvas: Canvas) {
        obstacle1?.draw(canvas)
        obstacle2?.draw(canvas)

        canvas.drawRect(0f, 0f, screenWidth.toFloat(), (screenHeight / 5).toFloat(), linearPaint)

        ground!!.draw(canvas)
        canvas.drawText(scoreStr + score, (screenWidth / 10).toFloat(), (screenHeight / 8).toFloat(),
            paint!!
        )
        if (gameOver) {
            gameover?.visibility = View.VISIBLE
            scoreText!!.text = score.toString() + ""
            return
        }

        if (gameStart && !gameOver) {
            ground?.step()
            rockets?.draw(canvas)
            obstacle1?.step()
            obstacle2?.step()

            if (rockets!!.pass(obstacle1!!, obstacle2!!, level)) {
                lastTime = System.currentTimeMillis()
                if (lastTime - firstTime >= 1500) {
                    score++
                }
                firstTime = lastTime
            }
            if (rockets!!.hit(obstacle1, obstacle2)) {
                gameStart = false
                gameOver = true
            }
        }

        if (!gameOver) rockets?.fly()


    }

    fun startGame() {
        start()
        gameStart = true
    }

    fun setOffset(offset: Float) {
        if (gameStart) {
            rockets?.setOffset(offset * magnification)
        }
    }
}